libbpf-tools/profile.c

```diff
@@ -307,53 +325,89 @@ static const char *usymname(unsigned long addr)
 	return sym ? sym->name : "[unknown]";
 }
 
-static void print_stacktrace(unsigned long *ip, symname_fn_t symname)
+static void print_stacktrace(unsigned long *ip, symname_fn_t symname, struct fmt_t *f)
+{
+	int i;
+
+	if (!f->folded) {
+		for (i = 0; ip[i] && i < env.perf_max_stack_depth; i++)
+			pr_format(symname(ip[i]), f);
+		return;
+	} else {
+		for (i = env.perf_max_stack_depth - 1; i >= 0; i--) {
+			if (!ip[i])
+				continue;
+
+			pr_format(symname(ip[i]), f);
+		}
+	}
+}
+
+static bool print_user_stacktrace(struct key_t *event, int stack_map,
+				  unsigned long *ip, struct fmt_t *f, bool delim)
+{
+	if (env.kernel_stacks_only || STACK_ID_EFAULT(event->user_stack_id))
+		return false;
+
+	if (delim)
+		pr_format(f->delim, f);
+
+	if (bpf_map_lookup_elem(stack_map, &event->user_stack_id, ip) != 0) {
+		pr_format("[Missed User Stack]", f);
+	} else {
+		syms = syms_cache__get_syms(syms_cache, event->pid);
+		if (!syms && f->folded)
+			fprintf(stderr, "failed to get syms\n");
+		else
+			print_stacktrace(ip, usymname, f);
+	}
+
+	return true;
+}
+
+static bool print_kern_stacktrace(struct key_t *event, int stack_map,
+				  unsigned long *ip, struct fmt_t *f, bool delim)
 {
-	for (size_t i = 0; ip[i] && i < env.perf_max_stack_depth; i++)
-		printf("    %s\n", symname(ip[i]));
+	if (env.user_stacks_only || STACK_ID_EFAULT(event->kern_stack_id))
+		return false;
+
+	if (delim)
+		pr_format(f->delim, f);
+
+	if (bpf_map_lookup_elem(stack_map, &event->kern_stack_id, ip) != 0)
+		pr_format("[Missed Kernel Stack]", f);
+	else
+		print_stacktrace(ip, ksymname, f);
+
+	return true;
 }
 
-static int print_count(struct key_t *event, __u64 count, int stack_map)
+static int print_count(struct key_t *event, __u64 count, int stack_map, bool folded)
 {
 	unsigned long *ip;
+	int ret;
+	struct fmt_t *fmt = &stacktrace_formats[folded];
 
 	ip = calloc(env.perf_max_stack_depth, sizeof(unsigned long));
 	if (!ip) {
 		fprintf(stderr, "failed to alloc ip\n");
 		return -ENOMEM;
 	}
 
-	/* kernel stack */
-	if (!env.user_stacks_only && !STACK_ID_EFAULT(event->kern_stack_id)) {
-		if (bpf_map_lookup_elem(stack_map, &event->kern_stack_id, ip) != 0)
-			printf("    [Missed Kernel Stack]\n");
-		else
-			print_stacktrace(ip, ksymname);
+	if (!folded) {
+		/* multi-line stack output */
+		ret = print_kern_stacktrace(event, stack_map, ip, fmt, false);
+		ret = print_user_stacktrace(event, stack_map, ip, fmt, ret && env.delimiter);
```

Comment:
The 'ret' is not used here.

Reply:
Yes. you are right.
It was added just to look similar to the line above. But it's useless. It has now been deleted.
Thank you very much for the detailed feedback.